runtime/drivers/blob/blobdownloader.go

```diff
@@ -138,176 +154,164 @@ func NewIterator(ctx context.Context, bucket *blob.Bucket, opts Options, l *zap.
-		return nil, err
+	files, ok := <-it.files
+	if !ok {
+		if it.err == nil {
+			// downloaded successfully
+			it.err = io.EOF
+		}
+		return nil, it.err
 	}
-
+	it.lastBatch = files
 	// clients can make changes to slice if passing the same slice that iterator holds
 	// creating a copy since we want to delete all these files on next batch/close
-	result := make([]string, len(it.localFiles))
-	copy(result, it.localFiles)
+	result := make([]string, len(it.lastBatch))
+	copy(result, it.lastBatch)
 	return result, nil
 }
 
-// NextBatch downloads next n files and copies to local directory
-func (it *blobIterator) NextBatch(n int) ([]string, error) {
-	if !it.HasNext() {
-		return nil, io.EOF
-	}
-	if len(it.nextPaths) != 0 {
-		paths := it.nextPaths
-		it.index += len(paths)
-		it.nextPaths = nil
-		return paths, nil
+func (it *blobIterator) Size(unit drivers.ProgressUnit) (int64, bool) {
+	switch unit {
+	case drivers.ProgressUnitByte:
+		var size int64
+		for _, obj := range it.objects {
+			if obj.full {
+				size += obj.obj.Size
+			} else {
+				// TODO :: make it more accurate considering more data can be downloaded
+				size += int64(obj.extractOption.limitInBytes)
+			}
+		}
+		return size, true
+	case drivers.ProgressUnitFile:
+		return int64(len(it.objects)), true
+	default:
+		return 0, false
 	}
+}
 
-	if !it.opts.KeepFilesUntilClose {
-		// delete files created in last iteration
-		fileutil.ForceRemoveFiles(it.localFiles)
-	}
-	start := it.index
-	end := it.index + n
-	if end > len(it.objects) {
-		end = len(it.objects)
-	}
-	it.index = end
+func (it *blobIterator) KeepFilesUntilClose(keepFilesUntilClose bool) {
+	it.opts.KeepFilesUntilClose = keepFilesUntilClose
+}
 
-	// new slice creation is not necessary on every iteration
-	// but there may be cases where n in first batch is different from n in next batch
-	// to keep things easy creating a new slice every time
-	it.localFiles = make([]string, end-start)
+func (it *blobIterator) init() {
 	g, grpCtx := errgroup.WithContext(it.ctx)
+	it.grp = g
 	g.SetLimit(_concurrentBlobDownloadLimit)
-	for i, obj := range it.objects[start:end] {
-		obj := obj
-		index := start + i // with repect to object slice
-		g.Go(func() error {
+	var stop bool
+	files := make([]string, 0)
+	var sizeInBytes int64
+	var wg sync.WaitGroup
+	for i := 0; i < len(it.objects) && !stop; i++ {
+		select {
+		case <-it.ctx.Done():
+			it.err = it.ctx.Err()
+			close(it.files)
+			return
+		default:
+			obj := it.objects[i]
+			// we can download partial file as well but its okay to use full object size since its okay to not have exact size
+			sizeInBytes += obj.obj.Size
 			// need to create file by maintaining same dir path as in glob for hivepartition support
 			filename := filepath.Join(it.tempDir, obj.obj.Key)
-			if err := os.MkdirAll(filepath.Dir(filename), os.ModePerm); err != nil {
-				return err
-			}
+			files = append(files, filename)
+			wg.Add(1)
+			g.Go(func() error {
+				defer wg.Done()
```

Comment:
I didn't understand calls to both `g.Go` and `wg.Wait()` happens in the same goroutine, that goroutine can either be waiting on `g.Go` when it can't spawn new goroutines or `wg.Wait()` when enough data has been downloaded (with download in progress for some).